File: model/src/main/scala/ru/mipt/acsl/decode/model/proxy/ExistingElementsProxyResolver.scala
package ru.mipt.acsl.decode.model.proxy

import ru.mipt.acsl.decode.model.Referenceable
import ru.mipt.acsl.decode.model.naming.Fqn
import ru.mipt.acsl.decode.model.proxy.path.{ProxyPath, TypeName}
import ru.mipt.acsl.decode.model.registry._

/**
  * Created by metadeus on 20.02.16.
  */
case object ExistingElementsProxyResolver extends DecodeProxyResolver {

  override def resolveElement(registry: Registry, path: ProxyPath): ResolvingResult[Referenceable] = {
    path match {
      case e: ProxyPath.FqnElement =>
        resolveFqnElement(registry, e)
      case l: ProxyPath.Literal =>
        resolveLiteralElement(registry, l)
      case _ => sys.error("not implemented")
    }
  }

  private def resolveLiteralElement(registry: Registry, literal: ProxyPath.Literal): ResolvingResult[Referenceable] = {
    val ns = registry.findNamespace(Fqn.DecodeNamespace).getOrElse(sys.error("decode namespace not found"))
    val elements = ns.aliases.filter(_.name == literal.mangledName)
    if (elements.size == 1) {
      val obj = elements.head.obj
      registry.resolve(obj)
      ResolvingResult(Some(obj))
    }
    else if (elements.size > 1)
      ResolvingResult(None, Result.error(s"must be exactly one element for $literal, found: ${elements.size}"))
    else
      ResolvingResult(None)
  }

  private def resolveFqnElement(registry: Registry, fqnElement: ProxyPath.FqnElement): ResolvingResult[Referenceable] = {
    val nsOption = registry.findNamespace(fqnElement.ns)
    if (nsOption.isEmpty)
      return ResolvingResult(None, Result.error(s"namespace not found ${fqnElement.ns.asMangledString}"))
    val ns = nsOption.get
    fqnElement.element match {
      case e: TypeName =>
        val name = e.typeName
        val elements = ns.aliases.filter(_.name == name)
        if (elements.size == 1) {
          val obj = elements.head.obj
          registry.resolve(obj)
          ResolvingResult(Some(obj))
        }
        else
          ResolvingResult(None, Result.error(s"must be exactly one element for $fqnElement, found: ${elements.size}"))
      case _ =>
        ResolvingResult(None)
    }
  }
}

File: model/src/main/java/ru/mipt/acsl/decode/model/proxy/ExistingElementsProxyResolver.java
package ru.mipt.acsl.decode.model.proxy;

import ru.mipt.acsl.decode.model.Message;
import ru.mipt.acsl.decode.model.Referenceable;
import ru.mipt.acsl.decode.model.naming.ElementName;
import ru.mipt.acsl.decode.model.naming.Fqn;
import ru.mipt.acsl.decode.model.naming.Namespace;
import ru.mipt.acsl.decode.model.proxy.path.ProxyElementName;
import ru.mipt.acsl.decode.model.proxy.path.ProxyPath;
import ru.mipt.acsl.decode.model.proxy.path.TypeName;
import ru.mipt.acsl.decode.model.registry.Registry;
import ru.mipt.acsl.decode.model.registry.RegistryUtils;
import ru.mipt.acsl.decode.model.types.Alias;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * @author Artem Shein
 */
public class ExistingElementsProxyResolver implements DecodeProxyResolver {

    public static ExistingElementsProxyResolver newInstance() {
        return new ExistingElementsProxyResolver();
    }

    @Override
    public ResolvingResult<Referenceable> resolveElement(Registry registry, ProxyPath path) {
        if (path instanceof ProxyPath.FqnElement)
            return resolveFqnElement(registry, (ProxyPath.FqnElement) path);
        else if (path instanceof ProxyPath.Literal)
            return resolveLiteralElement(registry, (ProxyPath.Literal) path);
        else
            throw new RuntimeException("not implemented");
    }

    private ResolvingResult<Referenceable> resolveLiteralElement(Registry registry, ProxyPath.Literal literal) {
        Namespace ns = registry.findNamespace(Fqn.DECODE_NAMESPACE)
                .orElseThrow(() -> new RuntimeException("system namespace must exist"));
        List<Alias> elements = ns.aliases().stream().filter(a -> a.name().equals(literal.mangledName())).collect(
                Collectors.toList());
        if (elements.size() == 1)
        {
            Alias alias = elements.get(0);
            RegistryUtils.resolve(registry, alias);
            Referenceable obj = alias.referenceable();
            return ResolvingResult.newInstance(obj, RegistryUtils.resolve(registry, obj));
        }
        else if (elements.size() > 1)
            return ResolvingResult.newInstance(
                    Message.newError(String.format("must be exactly one element for %s, found: %d", literal, elements.size())));
        else
            return ResolvingResult.newInstance();
    }

    private ResolvingResult<Referenceable> resolveFqnElement(Registry registry, ProxyPath.FqnElement fqnElement) {
        Optional<Namespace> nsOption = registry.findNamespace(fqnElement.ns());
        if (!nsOption.isPresent())
            return ResolvingResult.newInstance(
                    Message.newError(String.format("namespace not found %s", fqnElement.ns().mangledNameString())));
        Namespace ns = nsOption.get();
        ProxyElementName element = fqnElement.element();
        if (element instanceof TypeName)
        {
            ElementName name = ((TypeName) element).typeName();
            List<Alias> elements =
                    ns.aliases().stream().filter(a -> a.name().equals(name)).collect(Collectors.toList());
            if (elements.size() == 1)
            {
                Alias alias = elements.get(0);
                RegistryUtils.resolve(registry, alias);
                Referenceable obj = alias.referenceable();
                return ResolvingResult.newInstance(obj, RegistryUtils.resolve(registry, obj));
            }
            else
                return ResolvingResult.newInstance(
                        Message.newError(String.format("must be exactly one element for %s, found: %d", fqnElement,
                                elements.size())));
        }
        else
            return ResolvingResult.newInstance();
    }

    private ExistingElementsProxyResolver() {

    }
}
